v3_async_wip/v3_async_wip/mqtt_client.py

```diff
@@ -438,28 +615,191 @@ async def wait_for_disconnect():
+                    if rc not in expected_subscribe_rc:
+                        logger.warning("Unexpected rc {} from Paho .subscribe()".format(rc))
+                    raise MQTTError(rc)
+
+                # Establish a pending subscribe
+                sub_done = self._loop.create_future()
+                self._pending_subs[mid] = sub_done
+
+            logger.debug("Waiting for subscribe response for mid {}".format(mid))
+            await sub_done
+        except asyncio.CancelledError:
+            if mid:
+                logger.debug("Subscribe for mid {} was cancelled".format(mid))
+            else:
+                logger.debug("Subscribe was cancelled before mid was assigned")
+            raise
+        finally:
+            # Delete any pending operation (if it exists)
+            async with self._mid_tracker_lock:
+                if mid and mid in self._pending_subs:
+                    del self._pending_subs[mid]
+
+    async def unsubscribe(self, topic):
+        """
+        Unsubscribe from a topic on the MQTT broker.
+
+        :param str topic: a single string which is the subscription topic to unsubscribe from.
+
+        :raises: ValueError if topic is None or has zero string length.
+        :raises: MQTTError if there is an error subscribing
+        """
+        try:
+            mid = None
+            logger.debug("Attempting unsubscribe")
+            # Using this lock allows us to be sure that the ACK won't come in before the
+            # Future can be added to the pending dictionary
+            async with self._mid_tracker_lock:
+                (rc, mid) = await self._loop.run_in_executor(
+                    None, functools.partial(self._mqtt_client.unsubscribe, topic=topic)
+                )
+                rc_msg = mqtt.error_string(rc)
+                logger.debug("Unsubscribe returned rc {} - {}".format(rc, rc_msg))
+                if rc != mqtt.MQTT_ERR_SUCCESS:
+                    if rc not in expected_unsubscribe_rc:
+                        logger.warning("Unexpected rc {} from Paho .unsubscribe()".format(rc))
+                    raise MQTTError(rc)
+
+                # Establish a pending unsubscribe
+                unsub_done = self._loop.create_future()
+                self._pending_unsubs[mid] = unsub_done
+
+            logger.debug("Waiting for unsubscribe response for mid {}".format(mid))
+            await unsub_done
+        except asyncio.CancelledError:
+            if mid:
+                logger.debug("Unsubscribe for mid {} was cancelled".format(mid))
+            else:
+                logger.debug("Unsubscribe was cancelled before mid was assigned")
+            raise
+        finally:
+            # Delete any pending operation (if it exists)
+            async with self._mid_tracker_lock:
+                if mid and mid in self._pending_unsubs:
+                    del self._pending_unsubs[mid]
+
+    async def publish(self, topic, payload):
```

Comment:
Fun fact: If you call Paho's publish method with Qos=0, `on_publish` will still get called. It just gets called automatically by Paho instead of having to wait for the PUBACK.  

Maybe we should add a Qos parameter here while we're building things up. I've had partners ask to publish as Qos 0, and now might be a good time to add it.

Reply:
That sounds like it should make a more full QoS support less painful than expected, but nonthless I think that's best for a separate PR.